library/core/src/num/nonzero.rs

```diff
@@ -83,6 +88,79 @@ impl_zeroable_primitive!(NonZeroIsize(isize));
 #[rustc_diagnostic_item = "NonZero"]
 pub struct NonZero<T: ZeroablePrimitive>(T);
 
+impl<T> NonZero<T>
+where
+    T: ZeroablePrimitive + PartialEq,
+{
+    /// Creates a non-zero if the given value is not zero.
+    #[stable(feature = "nonzero", since = "1.28.0")]
+    #[rustc_const_stable(feature = "const_nonzero_int_methods", since = "1.47.0")]
+    #[must_use]
+    #[inline]
+    pub const fn new(n: T) -> Option<Self> {
+        if n == T::ZERO {
+            return None;
+        }
+
+        // SAFETY: We just checked that there's no `0`.
+        Some(unsafe { Self(n) })
```

Comment:
This should fix the const traits issue.

```suggestion
        unsafe { crate::mem::transmute_copy(&n) }
```

Reply:
```rust
match n {
    T::ZERO => None,
    _ => Some(unsafe { self(n) }),
}
```

may also work without requiring a PartialEq impl